design-documents/remote-invoker.md

```diff
@@ -0,0 +1,618 @@
+
+### Explanations
+#### Need for an abstraction around network calls
+Say we're writing a Proxy version of a ProductRepositoryInterface service using HTTP client and RESTful API.
+```
+class ProductRepoProxy implements ProductRepositoryInterface
+{
+
+....
+
+    public function save(ProductInterface $product): ProductInterface
+    {
+        $baseUrl = $this->config->get('product-node-url');
+        $request = new HttpRequest($baseUrl .'/V1/product');
+        $request->setHeader('X-Magento-User-Type', $this->userContext->getUserType());
+        $request->setHeader('X-Magento-User-Id', $this->userContext->getUserId());
+        $request->setBody(
+            $this->jsonSerializer->serialize(
+                ['product' => $this->objectSerializer->toArray($product, ProductInterface::class)]
+            )
+        );
+        $productCreated = null;
+        $this->transport->send($request)->then(function ($httpResponse) use (&$productCreated) {
+            $data = json_decode($httpResponse->getBody());
+            $productCreated = $this->objectSerializer->fromArray(ProductInterface::class, $data);
+        })->error(function ($exception) use ($product) {
+            if ($exception instanceof NetworkFailException) {
+                //Retrying
+                return $this->save($product);
+            }
+        })->wait();
+        
+        return $productCreated;
+    }
+    
+    public function delete(string $id): void
+    {
+        //Same stuff: using config, userContext, serializers etc.
+        ....
+    }
+
+....
+
+}
+```
+
+Clearly sending such requests involves a lot of logic and dependencies on other classes,
+and an abstraction (InvokerInterface) which will provide a simple way to pass arguments and remote source data
+to execute a remote service request.
+
+#### Exposing internal logic as an external Web API
+Let's say we have a hypothetical situation: customer places an order _(Checkout)_, but for each item ordered we have to update "ordered" count for related products _(Catalog)_
+```
+Class OrderManagement implements OrderManagementInterface
+{
+ 
+....
+ 
+public function place(OrderInterface $order): void
+{
+ 
+....
+ 
+    foreach ($order->getItems() as $product) {
+        //Calling a service from another domain (Catalog), maybe local, maybe remote
+        $this->productRepository->incrOrdederedCount($product->getId());
+    }
+ 
+....
+ 
+}
+```
+If the remote call to _ProductRepositoryInterface_ service was done via RESTful (or GraphQL) web API the _incrOrdederedCount_ would have
+to be exposed as Web API and could be requested via HTTP by anyone and independently from context which we cannot afford since we can't just increase the counter without an order happening so only another service responsible for orders may call _incrOrdederedCount_.
+ 
+If done via the Invoker the _incrOrdederedCount_ wouldn't be exposed to be requested by HTTP clients via web API and wouldn't show up on
```

Comment:
There is no difference because Invoker makes the method exposed to all services, if I understand the proposal correctly.
Keep in mind that client-facing Web APIs are not the same as service-facing (inter-network) Web APIs. Only limited list of Web APIs (explicitly declared) will be exposed to the client on the BFF side. All other APIs will be available for services only.

Reply:
With RPC style communication there would be no need to re-declare our web API endpoints and there would be a clear differentiation between web API endpoints for outside world and inner-network